<div align="center">
    <h1>Juno-assembly</h1>
    <br />
    <h2>Pipeline to process bacterial raw sequencing data up to de-novo assembly and the accompanying statistics.</h2>
    <br />
    <img src="https://via.placeholder.com/150" alt="pipeline logo">
</div>

## Pipeline information

* **Author(s):**            Alejandra Hernández Segura, Robert Verhagen and Ernst Hamer.
* **Organization:**         Rijksinstituut voor Volksgezondheid en Milieu (RIVM)
* **Department:**           Infektieziekteonderzoek, Diagnostiek en Laboratorium Surveillance (IDS), Bacteriologie (BPD)
* **Start date:**           01 - 04 - 2020
* **Commissioned by:**      Maaike van den Beld

## About this project

The goal of this pipeline is to generate assemblies from raw fastq files. The input of the pipeline is raw Illumina paired-end data (read length > 99 bp) in the form of two fastq files (with extension .fastq, .fastq.gz, .fq or .fq.gz), containing the forward and the reversed reads ('R1' and 'R2' must be part of the file name, respectively). On the basis of the generated genome assemblies, low quality and contaminated samples can be excluded for downstream analysis. __Note:__ The pipeline has been validated only in gastroenteric bacteria (_Salmonella_, _Shigella_, _Listeria_ and STEC) but it could theoretically be used in other genera/species.

The pipeline uses the following tools:
1. [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) (Andrews, 2010) is used to assess the quality of the raw Illumina reads
2. [Trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic) (Bolger, Lohse, & Usadel, 2014) is used to remove poor quality data and adapter sequences. The sliding window option of Trimmomatic starts scanning at the 5’ end and clips the read once the average quality within the window falls below a threshold. The sliding window is the number of nucleotides over which Trimmomatic calculates an average phred quality score, a measure of the quality of the identification of the nucleobases generated by automated DNA sequencing. The Trimmomatic minlen config parameter is set to 50, this parameter is used to drop the read if the read is below a specific length.
3. FastQC is used once more to assess the quality of the trimmed reads
4. [Picard](https://broadinstitute.github.io/picard/) determines the library fragment lengths
5. The reads are assembled into scaffolds by [SPAdes](https://cab.spbu.ru/software/spades/) (Bankevich et al., 2012) by means of _de novo_ assembly of the genome. SPAdes uses k-mers for building an initial de Bruijn graph and on following stages it performs graph-theoretical operations to assemble the genome. Kmer sizes of 21, 33, 55, 77 and 99 were used. For _de novo_ assembly, SPAdes isolate mode is used. 
6. [QUAST](http://quast.sourceforge.net/) (Gurevich, Saveliev, Vyahhi, & Tesler, 2013) is used to assess the quality of the filtered scaffolds. 
7. To assess the quality of the microbial genomes, [CheckM](https://ecogenomics.github.io/CheckM/) (Parks, Imelfort, Skennerton, Hugenholtz, & Tyson, 2015) is used. CheckM calculates scores for completeness, contamination and strain heterogeneity. 
8. [Bbtools](https://jgi.doe.gov/data-and-tools/bbtools/) (Bushnell, 2014) is used to generate scaffold alignment metrics. 
9. [MultiQC](https://multiqc.info/) (Ewels, Magnusson, Lundin, & Käller, 2016) is used to summarize analysis results and quality assessments in a single report for dynamic visualization.

![Image of pipeline](https://github.com/DennisSchmitz/BAC_gastro/blob/master/juno_pipeline.png)

## Prerequisities

* **Linux + conda** A Linux-like environment with at least 'miniconda' installed. 
* **Python3.7.6** .


## Installation

1. Clone the repository:

```
git clone https://github.com/RIVM-bioinformatics/Juno_pipeline.git
```
Alternatively, you can download it manually as a zip file (you will need to unzip it then).

2. Enter the directory with the pipeline and install the master environment:

```
cd Juno_pipeline
conda env install -f envs/master_env.yaml
```

## Parameters & Usage

### Command for help

* ```-h, --help``` Shows the help of the pipeline

### Required parameters

* ```-i, --input``` Directory with the input (fasta) files. The fasta files should be all in this directory (no subdirectories) and have the extension '.fastq', '.fastq.gz', '.fq' or '.fq.gz'. Important, any very small file (smaller than 3 kilobytes) will be ignored and not included in the pipeline results since they are probably empty files.  
* `--genus` Provide name of an approved genus to be used for CheckM. Please check --help-genera for a list of approved genera). Only one name is allowed, so if multiple samples are included in the input directory, it will be assumed that they all have the same genus. If different samples have different genus, then do not provide a `--genus` and use `--metadata` instead. If you prefer to avoid the hassle and just run the pipeline without needing the genus name, you can use the flag `no-checkm`. Keep in mind that the coverage of the assembly and other stats calculated by CheckM will not be obtained.

### Optional parameters

* `-m --metadata` Excel file (.xlsx) containing the information of the samples and their corresponding genus. It should contain at least a column called 'Monsternummer' containing the sample ID (same name than fastq files but removing the suffix _R1|2.fastq.gz) and another one called 'genus' containing the name of the genus. Mind the capital in the name 'Monsternummer'. Default is: /data/BioGrid/NGSlab/BAC_in_house_NGS/In- house_NGS_selectie_2021.xlsx (used at RIVM). An example metadata would be:

| __Monsternummer__ | __genus__ |
| :---: | :--- |
| sample1 | Salmonella |

*Note:* The fastq files corresponding to this sample would probably be something like sample1_S1_R1_0001.fastq.gz and sample1_S1_R2_0001.fastq.gz.

* `--no-checkm` If present, this flag tells the pipeline to not run CheckM or update the genus database from CheckM.
* ```-o --output``` Directory (if not existing it will be created) where the output of the pipeline will be collected. The default behavior is to create a folder called 'output' within the pipeline directory. 
* ```-c --cores```  Maximum number of cores to be used to run the pipeline. Defaults to 300 (it assumes you work in an HPC cluster).
* ```-l --local```  If this flag is present, the pipeline will be run locally (not attempting to send the jobs to a cluster). Keep in mind that if you use this flag, you also need to adjust the number of cores (for instance, to 2) to avoid crashes. The default is to assume that you are working on a cluster because the pipeline was developed in an environment where it is the case.
* ```-q --queue```  If you are running the pipeline in a cluster, you need to provide the name of the queue. It defaults to 'bio' (default queue at the RIVM). 
* ```-n --dryrun```, ```-u --unlock``` and ```--rerunincomplete``` are all parameters passed to Snakemake. If you want the explanation of these parameters, please refer to the [Snakemake documentation](https://snakemake.readthedocs.io/en/stable/).

### The base command to run this program. 

```
bash juno -i [path/to/input/dir] -o [path/to/output/dir] --genus [genus]
```

### An example on how to run the pipeline.

```
bash juno -i my_data -o my_results --genus Salmonella --local --cores 2
```

## Explanation of the output

* **log:** Log files with output and error files from each Snakemake rule/step that is performed. 
* **audit_trail:** Information about the versions of software and databases used.
* **output per sample:** The pipeline will create one subfolder per each step performed. These subfolders will in turn contain the results per sample and/or one with results combined for all samples (if applicable). To understand the ouptut, please refer to the manuals of each individual tool. The generated subfolders are:
    - clean_fastq: contains the fastq files after filtering low quality reads and trimming ends and/or adapters. This is what you would use in downstream analyses that require fastq files as input.
    - de_novo_assembly: results from the assembly step without filtering out the small contigs.
    - de_novo_assembly_filtered: results from the assembly step containing only contigs larger than 500bp. This is what you would normally use in downstream analyses that require assembly (fasta files) as input.
    - multiqc: MultiQC report for all samples run.
    - qc_clean_fastq: results of the quality control for fastq files. Run after filtering and trimming.
    - qc_de_novo_assembly: results of the quality control of the assemblies. Includes results of different tools (refer to those tools for interpretation).
    - qc_raw_fastq: results of the quality control for fastq files. Run before filtering and trimming. 
        
## Issues  

* All default values have been chosen to work with the RIVM Linux environment, therefore, there might not be applicable to other environments (although they should work if the appropriate arguments/parameters are given).
* Any issue can be reported in the [Issues section](https://github.com/RIVM-bioinformatics/Juno_pipeline/issues) of this repository.

## Future ideas for this pipeline

* Containerize.
* Improve the wrapper.
* Give the possibility to change more of the default parameters.

## License
This pipeline is licensed with an AGPL3 license. Detailed information can be found inside the 'LICENSE' file in this repository.

## Contact
* **Contact person:**       Alejandra Hernández Segura
* **Email**                 alejandra.hernandez.segura@rivm.nl
